### PR TITLE
Parameterized sql

### PR DESCRIPTION
**Pull request by agilliland**
### TODO
-  [x] "lorem ipsum" in Variables -> Help -> Syntax
-  [x] Frontend test failures

### CODE REVIEW

File: src/metabase/api/card.clj [hunk 1]
```diff
@@ -170,32 +165,25 @@
...
-                       (:archived card))       :card-unarchive
-                  :else                        :card-update)]
-      (events/publish-event event (assoc (Card id) :actor_id *current-user-id*)))))
+    (cond
+      ;; card is being archived
+      (and archived
+           (not (:archived card))) (card/archive-card id *current-user-id*)
+      ;; card is being unarchived
+      (and (not (nil? archived))
+           (not archived)
```
Comment (camsaul) [inline src/metabase/api/card.clj]: you can probably just check `(false? archived)` instead

File: src/metabase/models/card.clj [hunk 2]
```diff
@@ -45,6 +46,63 @@
   (db/cascade-delete! 'CardLabel :card_id id))
 
 
+;;; ## ---------------------------------------- PERSISTENCE FUNCTIONS ----------------------------------------
+
+
+(defn create-card
```
Comment (camsaul) [inline src/metabase/models/card.clj]: Hate to sound like a broken record but these should really end in a bang (`!`). See the rational in PR #2759

**Comment by camsaul** [comment]
SQL params is all working now for variables (`{{x}}` syntax as well as `[[... {{x}}]]` optional syntax). Still need to figure out this dimension situation

**Comment by camsaul** [comment]
@tlrobinson this is all g2g besides the frontend test failures & lorem ipsum. For test failures I'm seeing:

```
/home/ubuntu/metabase/frontend/src/metabase/components/Select.jsx
  47:48  warning  'searchProp' is missing in props validation             react/prop-types
  47:60  warning  'searchCaseInsensitive' is missing in props validation  react/prop-types

/home/ubuntu/metabase/frontend/src/metabase/dashboard/components/parameters/widgets/TextWidget.jsx
  29:27  warning  'className' is missing in props validation  react/prop-types
  29:38  warning  'isEditing' is missing in props validation  react/prop-types

/home/ubuntu/metabase/frontend/src/metabase/query_builder/template_tags/TagEditorParam.jsx
  53:22  warning  'databaseFields' is missing in props validation  react/prop-types

/home/ubuntu/metabase/frontend/src/metabase/query_builder/template_tags/TagEditorSidebar.jsx
  54:118  warning  'databaseFields' is missing in props validation  react/prop-types
  64:41   warning  'databaseFields' is missing in props validation  react/prop-types

/home/ubuntu/metabase/frontend/src/metabase/query_builder/template_tags/TagValuePicker.jsx
  3:8  warning  '_' is defined but never used  no-unused-vars

✖ 8 problems (0 errors, 8 warnings)
```

if you want to take a look at those?

**Comment by camsaul** [comment]
Also @mazameli or whoever let's put something in place of the lorem ipsum, I can put it there if we have something or if someone else wants to do it

**Comment by mazameli** [comment]
Yeah, the idea is that the panel should have a really quick summary of what this is for, and the syntax for using it, with a link to the full documentation. The syntax has changed a bit recently, so I feel less confident taking a stab at writing this. Might be a job for @salsakran.

Relatedly, the last time I checked out this branch locally, there was a mismatch between the tooltip for the button that brings out the side panel vs. the heading for the side panel. They should match, though now I'm not sure what we call this — Parameters?

**Comment by camsaul** [comment]
For reference this is the part with the lorem ipsum:

screen shot 2016-07-19 at 2 40 14 pm

**Comment by salsakran** [comment]
How about

```
{{VariableName}} creates a variable in this SQL template called "VariableName". You can give this variable a type. All variable types other than `dimension` will cause an input widget to be placed on this question. When this input widget is filled in, that value replaces the tag in the SQL template.
Example: 
`SELECT count(*) FROM events WHERE type={{event_type}}`
Creates a template that takes an event type as a parameter.

[[a clause with a {{VariableName}}]] creates an optional clause in the template. If VariableName is set, then the entire clause is placed into the template. If not, then the entire clause is ignored.

`SELECT count(*) FROM events [[ WHERE type={{event_type}}]]`
Creates a template that takes an event type as an optional parameter.

```

**Comment by mazameli** [comment]
Seems like a good overview, though it sort of omits what the `dimension` type does. Let's also add, "[Read the full documentation](#)" at the end.

**Comment by salsakran** [comment]
any better?

**Comment by tlrobinson** [comment]
This is looking good. One thing I noticed is we still have the migration for adding `parameters` to cards, but we don't actually need it since the card parameters are currently just generated on the fly from template_tags. Should we remove that migration?

**Comment by camsaul** [comment]
@tlrobinson where? I removed that migration

**Comment by tlrobinson** [comment]
Oh, I didn't realize you removed it. 👍

**Comment by mazameli** [comment]
Maybe this? 

{{VariableName}} creates a variable in this SQL template called "VariableName". Variables can be given types in the side panel, which changes their behavior. All variable types other than `dimension` will cause an input widget to be placed on this question. When this input widget is filled in, that value replaces the tag in the SQL template.

Giving a variable the `dimension` type allows you to link SQL cards to dashboard filters. A `dimension` variable is a template tag that inserts SQL snippets similar to those generated by our GUI when adding filters on existing columns. When adding a dimension, you should link that variable to a specific column. Dimension variable tags should be used inside of a WHERE clause. 

Example: 
`SELECT count(*) FROM events WHERE type={{event_type}}`
Creates a template that takes an event type as a parameter.

[[a clause with a {{VariableName}}]] creates an optional clause in the template. If `VariableName` is set, then the entire clause is placed into the template. If not, then the entire clause is ignored.

Example:
`SELECT count(*) FROM events [[ WHERE type={{event_type}} ]]`
Creates a template that takes an event type as an optional parameter.

[Read the full documentation](url)

**Comment by camsaul** [comment]
So all that we have left to do is stick the syntax explanation in the appropriate spot and @tlrobinson we have 2 frontend test failures if you can tackle those 😬 :

```
 Failed Tests:
card
createCard
should return a new card
Chrome 43.0.2357 (Linux 0.0.0)
1) Expected Object({ name: null, public_perms: 0, display: 'table', visualization_settings: Object({  }), dataset_query: Object({  }), parameters: [  ] }) to equal Object({ name: null, public_perms: 0, display: 'table', visualization_settings: Object({  }), dataset_query: Object({  }) }).
at Object.<anonymous> (/home/ubuntu/metabase/frontend/test/unit/lib/card.spec.js:55:46)

should set the name if supplied
Chrome 43.0.2357 (Linux 0.0.0)
2) Expected Object({ name: 'something', public_perms: 0, display: 'table', visualization_settings: Object({  }), dataset_query: Object({  }), parameters: [  ] }) to equal Object({ name: 'something', public_perms: 0, display: 'table', visualization_settings: Object({  }), dataset_query: Object({  }) }).
at Object.<anonymous> (/home/ubuntu/metabase/frontend/test/unit/lib/card.spec.js:65:57)


157 total155 passed2 failed0 skipped
```

**Comment by tlrobinson** [comment]
K I'll take care of those things. @mazameli do we have a style we'd like to use for the code snippets in the docs? Presumably a monospace font, perhaps a different color/background color?

**Comment by camsaul** [comment]
ya @tlrobinson use the background & font [from this website](http://www.warnerbros.com/archive/spacejam/movie/jam.htm)

**Comment by mazameli** [comment]
No, I don't think we do have a snippet style yet. How about this:

**text:**
font-family: Monaco;
font-size: 13px;
color: #8691AC;
line-height: 24px;   // incidentally, the `line-height` of all the paragraph text in the side bar could stand to be 24px; it's a bit tight right now

**background:**
background: #E9F2F5;
border-radius: 2px;
We'll probably have to eyeball how much padding around the text there is.

screen shot 2016-07-20 at 12 10 49 pm

**Comment by camsaul** [comment]
is this ready to merge?

**Comment by mazameli** [comment]
- Dropdowns need `cursor:pointer` 😡😡😡😡😡😡😡 😆  
- Not sure where the word "tag" got introduced, but I'd be in favor of removing it in place of existing terminology
- `Variable type` dropdown's placeholder text should probably just say `Select…`
- It would be cool if we could highlight the `variable type` dropdown in some way to indicate it's required. in the designs, i used a blue border, but that might be weird since we are using that for filters. though I was originally against this, we could also default the variable type to one of the options.
- Do we want to try and have friendlier error messages than `Assert failed: Unable to substitute 'prod_cat': param not specified. Found: null (contains? params k)`?
- If I specify a default value for a variable, I would have expected that to show up in filter widget. Otherwise it looks like the question isn't being filtered at all.

**Comment by tlrobinson** [comment]
@mazameli regarding the last point, this might be a little hard to explain, but the problem is the default value is an attribute of the `template_tag` rather than the `parameter` widget (which is implicitly generated from the `template_tag`). Even if the `parameter` itself has no value the `template_tag` still gets that default value because otherwise the query would be syntactically invalid.

I can set it up so the `parameter` widget displays the `template_tag`'s default if no other value has been set, which would be correct, however then when you click the "x" on the `parameter` widget nothing will happen. Should I just hide the "x" when the value matches the default value of the `template_tag`?

**Comment by mazameli** [comment]
Gotcha, that makes sense. Yeah, I think it feels more consistent to do what you proposed: set the `parameter` widget to display the `template_tag`'s default if no other value has been set, and hide the `X` on the widget in the event that the value matches the default of the `template_tag`.

**Comment by mazameli** [comment]
I think Sameer's two remaining items were:

1. New `Variables icon` (and let's change the tooltip on the icon to say **Variables** instead of **Parameters**)
2. Make the `Run query` button appear when you input text into a text filter widget

This is what I came up with for a different Variables icon:
screen shot 2016-07-21 at 10 23 31 am

The path for the `x` icon svg is:
```
M32,3.85760518 C32,5.35923081 31.5210404,6.55447236 30.5631068,7.4433657 C29.6051732,8.33225903 28.4358214,8.77669903 27.0550162,8.77669903 C26.2265331,8.77669903 25.4110072,8.67314019 24.6084142,8.46601942 C23.8058212,8.25889864 23.111114,8.05178097 22.5242718,7.84466019 C22.2481108,8.03452091 21.8425054,8.44875625 21.3074434,9.08737864 C20.7723814,9.72600104 20.1682882,10.5026923 19.4951456,11.4174757 C20.116508,14.0582656 20.6170423,15.9352695 20.9967638,17.0485437 C21.3764852,18.1618179 21.7389411,19.2880202 22.0841424,20.4271845 C22.3775635,21.3419679 22.8090586,22.0582498 23.3786408,22.5760518 C23.9482229,23.0938537 24.8457328,23.3527508 26.0711974,23.3527508 C26.5199591,23.3527508 27.0809028,23.2664518 27.7540453,23.0938511 C28.4271878,22.9212505 28.9795016,22.7486524 29.4110032,22.5760518 L28.8414239,24.9061489 C27.1326775,25.6310716 25.6397043,26.1574957 24.3624595,26.4854369 C23.0852148,26.8133781 21.9460676,26.9773463 20.9449838,26.9773463 C20.2200611,26.9773463 19.5037792,26.9083071 18.7961165,26.7702265 C18.0884539,26.632146 17.4412111,26.3818788 16.8543689,26.0194175 C16.2157465,25.6396961 15.6763776,25.1650514 15.236246,24.5954693 C14.7961143,24.0258871 14.4207135,23.2319361 14.1100324,22.2135922 C13.9029116,21.5749698 13.7130537,20.850058 13.5404531,20.038835 C13.3678524,19.2276119 13.1952544,18.51133 13.0226537,17.8899676 C12.5221118,18.6321504 12.1596559,19.1844642 11.9352751,19.5469256 C11.7108942,19.9093869 11.3829579,20.4185512 10.9514563,21.0744337 C9.5879112,23.1629015 8.4056145,24.6515597 7.40453074,25.5404531 C6.40344699,26.4293464 5.20389049,26.8737864 3.80582524,26.8737864 C2.75296129,26.8737864 1.85545139,26.5199604 1.11326861,25.8122977 C0.371085825,25.1046351 0,24.1812355 0,23.0420712 C0,21.5059254 0.478959612,20.2934241 1.4368932,19.4045307 C2.3948268,18.5156374 3.56417864,18.0711974 4.94498382,18.0711974 C5.77346693,18.0711974 6.56741799,18.1704413 7.32686084,18.368932 C8.08630369,18.5674228 8.80258563,18.7874853 9.47572816,19.0291262 C9.73462913,18.8220054 10.1359196,18.4164 10.6796117,17.8122977 C11.2233037,17.2081955 11.814452,16.4573939 12.4530744,15.5598706 C11.8834923,13.2470219 11.4174775,11.5037815 11.0550162,10.3300971 C10.6925548,9.15641269 10.321469,7.99137579 9.94174757,6.83495146 C9.63106641,5.90290796 9.18231146,5.18231107 8.59546926,4.67313916 C8.00862706,4.16396725 7.12837696,3.90938511 5.95469256,3.90938511 C5.43689061,3.90938511 4.85868712,3.99999909 4.22006472,4.18122977 C3.58144233,4.36246045 3.04638835,4.53074356 2.61488673,4.68608414 L3.18446602,2.35598706 C4.73787184,1.66558447 6.20927029,1.14779029 7.5987055,0.802588997 C8.98814071,0.457387702 10.1488627,0.284789644 11.0809061,0.284789644 C11.9266493,0.284789644 12.6515612,0.345198964 13.2556634,0.466019417 C13.8597657,0.586839871 14.4983785,0.845736958 15.171521,1.24271845 C15.7928834,1.62243987 16.3322523,2.10139948 16.789644,2.67961165 C17.2470357,3.25782382 17.6224365,4.04745994 17.9158576,5.04854369 C18.1229784,5.73894628 18.3128362,6.45522822 18.4854369,7.197411 C18.6580375,7.93959379 18.8047459,8.5782066 18.9255663,9.11326861 C19.2880277,8.56094654 19.6634285,7.99137294 20.0517799,7.40453074 C20.4401314,6.81768854 20.7723827,6.29989437 21.0485437,5.85113269 C22.3775687,3.76266485 23.5684953,2.2653767 24.6213592,1.3592233 C25.6742232,0.453069903 26.8651498,0 28.1941748,0 C29.2815588,0 30.1876986,0.358140971 30.9126214,1.07443366 C31.6375441,1.79072634 32,2.71844091 32,3.85760518 L32,3.85760518 Z
```

**Pull request merged**